external/storm-kafka-client/src/main/java/org/apache/storm/kafka/spout/KafkaSpoutRetryExponentialBackoff.java

```diff
@@ -54,59 +56,92 @@
     private static class RetryEntryTimeStampComparator implements Serializable, Comparator<RetrySchedule> {
         @Override
         public int compare(RetrySchedule entry1, RetrySchedule entry2) {
-            return Long.valueOf(entry1.nextRetryTimeNanos()).compareTo(entry2.nextRetryTimeNanos());
+            int result = Long.valueOf(entry1.nextRetryTimeMs()).compareTo(entry2.nextRetryTimeMs());
+            
+            if(result == 0) {
+                //TreeSet uses compareTo instead of equals() for the Set contract
+                //Ensure that we can save two retry schedules with the same timestamp
+                result = entry1.hashCode() - entry2.hashCode();
+            }
+            return result;
         }
     }
 
     private class RetrySchedule {
         private final KafkaSpoutMessageId msgId;
-        private long nextRetryTimeNanos;
+        private long nextRetryTimeMs;
 
         public RetrySchedule(KafkaSpoutMessageId msgId, long nextRetryTime) {
             this.msgId = msgId;
-            this.nextRetryTimeNanos = nextRetryTime;
+            this.nextRetryTimeMs = nextRetryTime;
             LOG.debug("Created {}", this);
         }
 
         public void setNextRetryTime() {
-            nextRetryTimeNanos = nextTime(msgId);
+            nextRetryTimeMs = nextTime(msgId);
             LOG.debug("Updated {}", this);
         }
 
-        public boolean retry(long currentTimeNanos) {
-            return nextRetryTimeNanos <= currentTimeNanos;
+        public boolean retry(long currentTimeMs) {
+            return nextRetryTimeMs <= currentTimeMs;
         }
 
         @Override
         public String toString() {
             return "RetrySchedule{" +
                     "msgId=" + msgId +
-                    ", nextRetryTime=" + nextRetryTimeNanos +
+                    ", nextRetryTimeMs=" + nextRetryTimeMs +
                     '}';
         }
 
+        @Override
+        public int hashCode() {
```

Comment:
Classes of this type are put in TreeSet. If you add a value to TreeSet that equals an object already in the set with respect to compareTo, the new value is dropped. This is a problem in one of the tests I added that uses time simulation, because with simulated time the clock is stopped (i.e. the schedules have the same timestamps). I suppose it could occur outside tests, but it would be extremely rare. No reason not to fix it though.